src/main/java/com/ably/kafka/connect/ChannelSinkTask.java

```diff
@@ -61,16 +62,21 @@ public void put(Collection<SinkRecord> records) {
         }
     }
 
-    private void publishSingleRecord(SinkRecord record) {
-        SinkRecord suspendRecord = null;
-        if (suspended.get()){
-            suspendQueue.enqueue(record);
-        } else if ((suspendRecord = suspendQueue.dequeue()) != null) {
-            while (suspendRecord != null && !suspended.get()){
+    public void onSuspendedStateChange(boolean suspended) {
+        this.suspended.set(suspended);
```

Comment:
We might need some extra synchronisation here - in theory we could get a situation where:

- Channel is suspended, with records in the queue
- Channel enters non-suspended state (called on a separate thread by ably-java) and starts dequeueing records.
- `put` is called and then publishes some records in the middle of dequeuing the suspended ones.

So I think we might need to synchronise whilst records are being published, to ensure that we don't get random ones popping up in the middle?

Reply:
I changed the implementation so that new messages are just added to the queue if we're currently dequeueing messages [e4294d8](https://github.com/ably/kafka-connect-ably/pull/108/commits/e4294d81b9a0bcfbf7a69f691b8e05d74bb20a64)

Reply:
Just to summarise our offline chats - I was thinking that we need more synchronization, probably against the `suspendQueue`, whenever we are dequeuing or publishing a new message, to make sure that messages don't get lost due to race conditions.

Reply:
I've added a commit with your suggestion @AndyTWF, let me know if it all looks good to you now?

Reply:
Slight mistake on my part - I think the publishFrom in the `onSuspendedStateChange` block needs to move up a line to be within the `synchronized` block to ensure that the message is actually published before allowing the main `put` method to do a publish

Reply:
Ah good point - I've made that change now https://github.com/ably/kafka-connect-ably/pull/108/commits/d2b601965fba56e2c78bb497a6ea6b595718086b

Reply:
Resolving this on behalf of Andy as I see he has approved.